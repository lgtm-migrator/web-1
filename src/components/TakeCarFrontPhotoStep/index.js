import React, { useEffect } from "react"
import Button from "@material-ui/core/Button"
import useUserMedia from "../../hooks/useUserMedia"
import useVideoRef from "../../hooks/useVideoRef"
import useCanvasRef from "../../hooks/useCanvasRef"
import Skeleton from "@material-ui/lab/Skeleton"
import Card from "@material-ui/core/Card"
import CardContent from "@material-ui/core/CardContent"
import CardActions from "@material-ui/core/CardActions"
import Fab from "@material-ui/core/Fab"
import CameraAltIcon from "@material-ui/icons/CameraAlt"

const TakeCarFrontPhotoStep = () => {
  const constraints = {
    audio: false,
    video: {
      facingMode: "environment",
      height: { min: 360, ideal: 720 },
      width: { min: 480, ideal: 1280 },
    },
  }

  const [stream, getUserMedia] = useUserMedia(constraints)

  /**
   * Create video ref using hook
   */
  const [videoRef, videoDimensions] = useVideoRef()

  /**
   * Create canvas ref using hook
   */
  const canvasRef = useCanvasRef()

  /**
   * Send this stream to canvas
   */
  const handleClick = () => {
    console.log("Capturing photo...")
    canvasRef.current
      .getContext("2d")
      .drawImage(
        videoRef.current,
        0,
        0,
        videoDimensions.width,
        videoDimensions.height
      )
  }

  const handleAskCameraPermission = () => {
    console.log("Asking camera permission....")
    getUserMedia()
  }

  /**
   * Get canvas width and height from video
   */
  useEffect(() => {
    if (videoDimensions.width > 0 && videoDimensions.height > 0) {
      console.log("Updating video dimensions...", videoDimensions)
      canvasRef.current.width = videoDimensions.width
      canvasRef.current.height = videoDimensions.height
    }
  }, [canvasRef, videoDimensions])

  /**
   * Set this stream to video source
   */
  useEffect(() => {
    videoRef.current.srcObject = stream
  }, [stream, videoRef])

  return (
    <>
      {!stream && (
        <Card>
          <CardContent>Por favor, libere acesso à sua câmera</CardContent>
          <CardActions>
            <Button onClick={handleAskCameraPermission}>Usar câmera</Button>
          </CardActions>
        </Card>
      )}

      <div style={{ position: "relative" }}>
        <video
          autoPlay={true}
          controls={false}
          controlsList={"nodownload"}
          disablePictureInPicture={true}
          muted={true}
          playsInline={true}
          ref={videoRef}
          style={{ maxWidth: "100%", height: "auto" }}
        >
          <track />
        </video>
        <Fab
          onClick={handleClick}
          style={{
            position: "absolute",
            bottom: "0px",
            left: "50%",
            transform: "translate(-50%, 50%)",
          }}
          variant={"extended"}
        >
          <CameraAltIcon />
          &nbsp;Tirar foto
        </Fab>
      </div>

      <Skeleton style={{ maxWidth: "100%", height: "auto" }} variant="rect" />

      <canvas ref={canvasRef} style={{ maxWidth: "100%", height: "auto" }} />
    </>
  )
}

export default TakeCarFrontPhotoStep
