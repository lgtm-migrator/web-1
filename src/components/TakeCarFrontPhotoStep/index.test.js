import React from "react"
import renderer from "react-test-renderer"
import TakeCarFrontPhotoStep from "."

describe("Take car front photo step", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<TakeCarFrontPhotoStep />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
