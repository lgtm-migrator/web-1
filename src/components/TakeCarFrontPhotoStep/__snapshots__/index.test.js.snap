// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Take car front photo step should render correctly 1`] = `
Array [
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardContent-root"
    >
      Por favor, libere acesso à sua câmera
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Usar câmera
        </span>
      </button>
    </div>
  </div>,
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <video
      autoPlay={true}
      controls={false}
      controlsList="nodownload"
      disablePictureInPicture={true}
      muted={true}
      playsInline={true}
      style={
        Object {
          "height": "auto",
          "maxWidth": "100%",
        }
      }
    >
      <track />
    </video>
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-extended"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "bottom": "0px",
          "left": "50%",
          "position": "absolute",
          "transform": "translate(-50%, 50%)",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <circle
            cx="12"
            cy="12"
            r="3.2"
          />
          <path
            d="M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
          />
        </svg>
         Tirar foto
      </span>
    </button>
  </div>,
  <span
    className="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
    style={
      Object {
        "height": "auto",
        "maxWidth": "100%",
        "width": undefined,
      }
    }
  />,
  <canvas
    aria-label="Prévia da foto que você acabou de tirar"
    style={
      Object {
        "height": "auto",
        "maxWidth": "100%",
      }
    }
  />,
]
`;
